packages/server-wallet/src/onchain-service/onchain-service.ts

```diff
@@ -0,0 +1,297 @@
+import {ContractArtifacts} from '@statechannels/nitro-protocol';
+import {Bytes32, Address} from '@statechannels/client-api-schema';
+import {providers, Contract, BigNumber, Event, constants} from 'ethers';
+import {Evt} from 'evt';
+
+import {
+  MinimalTransaction,
+  OnchainServiceInterface,
+  AssetHolderInformation,
+  OnchainServiceConfiguration,
+  TransactionSubmissionServiceInterface,
+  DEFAULT_MAX_TRANSACTION_ATTEMPTS,
+  ChannelEventRecord,
+  DepositedEvent,
+  EvtContainer,
+  ContractEvent,
+  ChannelEventRecordMap,
+} from './types';
+
+// FIXME: replace with
+// import {Wallet as ChannelWallet, WalletError as ChannelWalletError} from '@statechannels/server-wallet';
+import {Wallet as ChannelWallet, WalletError as ChannelWalletError} from '..';
+
+type Values<E> = E[keyof E];
+
+const getAssetHolderInformation = async (
+  assetHolderAddress: Address,
+  provider: providers.JsonRpcProvider
+): Promise<AssetHolderInformation> => {
+  // Even though we do not know if this is an ETH or ERC20 asset holder,
+  // we must instantiate the contract with the Erc20AssetHolderArtifact
+  // because it is the only one with the `Token` property within the abi
+  const assetHolder = new Contract(
+    assetHolderAddress,
+    ContractArtifacts.Erc20AssetHolderArtifact.abi,
+    provider
+  );
+  let tokenAddress: string;
+  try {
+    tokenAddress = await assetHolder.Token();
+  } catch (e) {
+    // Failed to get property, means this is an ETH asset holder
+    tokenAddress = constants.AddressZero;
+  }
+  return {tokenAddress, assetHolderAddress};
+};
+
+class OnchainServiceError extends ChannelWalletError {
+  readonly type = ChannelWalletError.errors.OnchainError;
+
+  // Reasons an error is thrown from the transaction submission
+  // service
+  static readonly reasons = {
+    notRegistered: 'Must call register channel',
+    noChannelWallet: 'Must attach channel wallet',
+  } as const;
+  constructor(
+    reason: Values<typeof OnchainServiceError.reasons>,
+    public readonly data: any = undefined
+  ) {
+    super(reason);
+  }
+}
+
+export class OnchainService implements OnchainServiceInterface {
+  private provider: providers.JsonRpcProvider;
+  private attempts: number;
+
+  // NOTE: V0 has an in-memory store, which will eventually be
+  // shifted to more permanent storage
+  private store: Map<string, ChannelEventRecord> = new Map();
```

Comment:
You may want to consider making this "store" an external component with an in-memory implementation to start. Then you can add methods for storing events, transaction IDs (for in progress transactions), private keys, and things like this.